src/cow.jl

```diff
@@ -0,0 +1,141 @@
+using IRTools
+using MacroTools
+
+using IRTools: BasicBlock, stmt, var
+
+const MUTATING_OPS = (:push!, :pop!, :setindex!)
+
+function copy_object(obj)
+    ct = _current_task()
+    ct.storage == nothing && (ct.storage = IdDict())
+    if !haskey(ct.storage, :_cow_asset)
+        ct.storage[:_cow_asset] = IdDict{Any, Bool}()
+    end
+    if !haskey(ct.storage, :_cow_copylog)
+        ct.storage[:_cow_copylog] = Dict{UInt64, Any}()
+    end
+
+    haskey(ct.storage[:_cow_asset], obj) && (return obj) # own this obj
+    if haskey(ct.storage[:_cow_copylog], objectid(obj)) # had copied this obj
+        ret = ct.storage[:_cow_copylog][objectid(obj)]
+        ct.storage[:_cow_asset][ret] = true
+        return ret
+    end
+
+    new_obj = deepcopy(obj)
+    ct.storage[:_cow_asset][new_obj] = true
+    ct.storage[:_cow_copylog][objectid(obj)] = new_obj
+    return new_obj
+end
+
+maybe_copy(x) = x
+maybe_copy(x::AbstractArray{T}) where {T<:Number} = copy_object(x)
```

Comment:
I guess we don't need the static type parameter here?
```suggestion
maybe_copy(x::AbstractArray{<:Number}) = copy_object(x)
```